### PR TITLE
Introduces HDF5 support and refactors training to use datasets

### PR DESCRIPTION
**Pull request by josegcpa**
This is a fairly large PR (sorry!) with the following changes:

1. Introduces a `npy`-to-`hdf5` pipeline (in `paper_experiments/dataset_to_hdf5.py`). This does all of the Cellpose heavy lifting relating to flow calculations and prepares an HDF5 file with two datasets: one with images (RGB) and the other with 5 channels (cell, cell type, instance, flow in X direction, flow in Y direction).
2. Introduces a class which handles the HDF5 as a PyTorch-compatible dataset (`ClassposeHDF5Dataset`)
3. Alters the `ClassposeTrainingDataset` to have an interface which is consistent with `ClassposeHDF5Dataset`
4. Introduces changes to `train.train_class_seg` so that it uses as input `torch` Datasets rather than the current messy interface which features files and arrays simultaneously
5. Introduces a separate module (`train_utils`) containing a number of training utilities in a bit more neatly packaged functions

This effectively achieves two things:
1. Decouples dataset generation from the training script, making things **considerably** simpler if/when we use different dataset formats
2. Prepares Classpose for training with massive datasets which cannot be effectively loaded into memory.

The tests are working and so is an _ad hoc_ test designed for this:

```bash
#!/bin/bash
set -e

export COMPILE_MODEL=1

# Define paths
CONIC_DATA_DIR="data/datasets/conic/train"
SMALL_CONIC_DATA_DIR="data/datasets/conic_small/train"
SMALL_CONIC_HDF5_FILE="data/datasets/conic_small/train.h5"
GLYSAC_DATA_DIR="data/datasets/glysac/train_multi_shape"

# Common training arguments derived from run-all.sh
BOILERPLATE="
    --batch_size 4 \
    --oversampling_method custom \
    --device cuda \
    --augmentation_strategy enhanced \
    --num_workers 4 \
    --use_uncertainty_weighting \
    --min_train_masks 0 \
    --train_fraction 1.0
"
EPOCHS=5

echo "0. Subset Conic to 100 samples ==="
mkdir -p $SMALL_CONIC_DATA_DIR
uv run python -c "
import os
import numpy as np

images_file = os.path.join('$CONIC_DATA_DIR', 'images.npy')
labels_file = os.path.join('$CONIC_DATA_DIR', 'labels.npy')

images = np.load(images_file)[:100]
labels = np.load(labels_file)[:100]

np.save(os.path.join('$SMALL_CONIC_DATA_DIR', 'images.npy'), images)
np.save(os.path.join('$SMALL_CONIC_DATA_DIR', 'labels.npy'), labels)
"

echo "1. Generating HDF5 dataset for Conic ==="
if [ ! -f $SMALL_CONIC_HDF5_FILE ]; then
    uv run python paper_experiments/dataset_to_hdf5.py \
        --data_dirs $SMALL_CONIC_DATA_DIR \
        --output_file $SMALL_CONIC_HDF5_FILE
fi

echo "2. Training model using NPY dataset ==="
uv run python paper_experiments/run_training.py \
    --data_path $SMALL_CONIC_DATA_DIR \
    --epochs $EPOCHS \
    --output_dir "models/conic_npy" \
    --model_name "full" $BOILERPLATE

echo "3. Training model using HDF5 dataset ==="
uv run python paper_experiments/run_training.py \
    --data_path $SMALL_CONIC_HDF5_FILE \
    --epochs $EPOCHS \
    --output_dir "models/conic_hdf5" \
    --model_name "full" $BOILERPLATE

echo "4. Training GlySAC model using npy dataset ==="
uv run python paper_experiments/run_training.py \
    --data_path $GLYSAC_DATA_DIR \
    --epochs $EPOCHS \
    --output_dir "models/glysac_array" \
    --model_name "full" $BOILERPLATE
```

**EDIT:** finished reviewing things from my side, now we are ready for the review 🤓

### CODE REVIEW

File: paper_experiments/dataset_to_hdf5.py [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+            f["labels"].resize(
+                (f["labels"].shape[0] + label_data.shape[0], *lbl_shape)
+            )
+            f["labels"][-label_data.shape[0] :] = label_data
+
+    logger.info("Done!")
+
+
+if __name__ == "__main__":
+    main()
```
Comment (sohmandal) [inline paper_experiments/dataset_to_hdf5.py]: Nice, so I undestand that this refactoring improves the batch-loading path because samples are read lazily during iteration, but it does not make startup memory-safe. Before the first batch, training still performs dataset-wide label work to derive `n_classes, oversampling probabilities, and class weights`. In single-GPU training, that means classpose still pays a potentially large one-time memorycost, which can be a hard blocker if the dataset itself is too large to fit in host memory. In ddp training, the same startup work is repeated independently on every rank, so the host-memory and startup-time problem is multiplied by world size. Can we avoid this by moving those startup statistics to a memory-safe path like either precompute and persist them during dataset conversion?

Comment (sohmandal) [reply paper_experiments/dataset_to_hdf5.py]: Alternatively we could compute them in a streaming way directly from HDF5 without ever building a full in-memory label list. Then rank-0 compute plus broadcast that. But I would say the previous option is probably more straight-forward

File: src/classpose/dataset.py [hunk 2]
```diff
@@ -103,6 +238,204 @@ def __getitem__(self, index: int) -> tuple[np.ndarray, np.ndarray]:
...
+        Fetch items from the HDF5 file by index or slice.
+
+        Args:
+            i (int | list[int] | slice): The index or slice to fetch.
+
+        Returns:
+            tuple[np.ndarray, np.ndarray]: The fetched images and labels.
+        """
+        if isinstance(i, slice):
+            i = [idx for idx in range(slice.start, slice.stop, slice.step or 1)]
```
Comment (sohmandal) [inline src/classpose/dataset.py]:
```suggestion
            i = [idx for idx in range(i.start, i.stop, i.step or 1)]
```

**Review by sohmandal — CHANGES_REQUESTED** [review]
This is a neat refactor!!! Can I please request the commented changes to make HDF5-based training fully lazy to avoid the memory scaling issues? There are some bugs around subset handling, etc. which happens in specific use-cases! But we can merge this PR without resolving those for now, if we do not plan on running the training in those specific contexts!

**Review by sohmandal — COMMENTED** [review]
New changes look fantastic and the pipeline is robust! All tests pass as well! This PR should be good to merge with the additional warning flag for class labels mismatch between train and validation!
